src/Npgsql/NpgsqlCommand.cs

```diff
@@ -1155,7 +1163,9 @@ internal async ValueTask<NpgsqlDataReader> ExecuteReader(CommandBehavior behavio
 
             try
             {
-                if (conn.TryGetBoundConnector(out var connector))
+                var connector = _connector;
+                Debug.Assert(connector is null ^ conn is null);
```

Comment:
Instead of having two modes - with connection or with connector - can we set _connector even when using (regular) connection mode? That would mean that we can always just look at _connector directly and not care about the connection (except for very specific things).

Reply:
There is one problem - at which point do we set it? We cannot do it whenever we set the NpgsqlConnection - it might still be closed at this point, so there will be no bound connector.

Reply:
We could set it at the very start of ExecuteReader, for example - it would simplify checks/state afterwards. But I'm definitely OK with not doing it, up to you.